src/core/dimensions.jl

```diff
@@ -144,7 +140,57 @@ function dim_names(ccd::AbstractCompositeComponentDef)
     return collect(dims)
 end
 
+"""
+    _check_time_redefinition(obj::AbstractCompositeComponentDef, keys::Union{Int, Vector, Tuple, AbstractRange})
+
+Run through all necesssary safety checks for redefining `obj`'s time dimenson to 
+a new dimension with keys `keys`.
+"""
+function _check_time_redefinition(obj::AbstractCompositeComponentDef, keys::Union{Int, Vector, Tuple, AbstractRange}) where T
+
+    # get useful variables 
+    curr_keys = time_labels(obj)
+    curr_first = obj.first
+    curr_last = obj.last
+    
+    new_keys = [keys...]
+    new_first = first(new_keys)
+    new_last = last(new_keys)
+
+    # (1) check that the shift is legal
+    if isa(obj, ModelDef)
+        new_first > curr_first && error("Cannot redefine the time dimension to start at $new_first because it is after the model's current start $curr_first.") 
+        curr_first > new_last && error("Cannot redefine the time dimension to end at $new_last because it is before the model's current start $curr_first")
+    else
+        new_first > curr_first && error("Cannot redefine the time dimension to start at $new_first, because it is after component $(nameof(obj))'s start $curr_first.") 
+        curr_first > new_last && error("Cannot redefine the time dimension to end at $new_last because it is before the component $(nameof(obj))'s current start $curr_first")
+    end
+    
+    # (2) check first and last
+    !(curr_first in new_keys) && error("The current first index ($curr_first) must exist within the model's new time dimension $new_keys.") # can be assumed since we cannot move the time forward
+    curr_last >= new_last && !(new_last in curr_keys) && error("The new last index ($new_last) must exist within the model's current time dimension $curr_keys, since the time redefinition contracts to an earlier year.")
+    curr_last < new_last && !(curr_last in new_keys) && error("The current last index ($curr_last) must exist within the model's redefined time dimension $new_keys, since the time redefinition expands to a later year.")
+
+    # (3) check that the overlap region between the current keys and new keys holds same keys
+    if length(curr_keys) > 1 && length(new_keys) > 1
+        if isuniform(curr_keys) # fixed timesteps
+            step_size(curr_keys) != step_size(new_keys) && error("Cannot redefine the time dimension to have a timestep size of $(step_size(new_keys)), must match the timestep size of current time dimension, $(step_size(curr_keys))")
+        
+        else # variable timesteps      
+            start_idx = 1 # can be assumed since we cannot move the time forward
+            new_last < curr_last ? end_idx = findfirst(isequal(curr_last), new_keys) : end_idx = length(curr_keys)
```

Comment:
Should the first return value instead be "findfirst(isequal(new_last), curr_keys)"? It looks like we're looking for how much of curr_keys ought to be retained in the case where the new values end before the current values. Which would be the portion of it that goes until the new ending.